src/NHibernate/Type/TypeFactory.cs

```diff
@@ -683,14 +703,19 @@ private static NullableType GetType(NullableType defaultUnqualifiedType, byte pr
 		/// </para>
 		/// </remarks>
 		public static NullableType GetSerializableType(System.Type serializableType)
+		{
+			return (NullableType) GetSerializedOrBasicType(serializableType);
+		}
+
+		private static IType GetSerializedOrBasicType(System.Type serializableType)
 		{
 			var key = serializableType.AssemblyQualifiedName;
 
 			// The value factory may be run concurrently, but only one resulting value will be yielded to all threads.
 			// So we should add the type with its other key in a later operation in order to ensure we cache the same
 			// instance for both keys.
 			var added = false;
-			var type = (NullableType)typeByTypeOfName.GetOrAdd(
+			var type = typeByTypeOfName.GetOrAdd(
```

Comment:
We have a double caching here, first by .Net type `AssemblyQualifiedName`, then by NHibernate type `Name`. If the type has just been cached by its `AssemblyQualifiedName`, it is not supposed to be already cached by it NHibernate `Name`, nor another type is supposed to be cached by the same NHibernate `Name`.

This is not a throw I have added by the way. It was already there, in former `AddToTypeOfName` method, removed by a46b2d5.

The others `GetSerializableType` are specialized by length and are not "double" cached.

Reply:
From what I see  NHibernate types registration is going inside static ctor in `TypeFactory` - there should not be racing conditions. Wouldn't it be enough to check that returned type is exactly `SerializableType` and not some other built-in type?
`type.GetType() == typeof(SerializableType)`

As it is my understanding that possible multiple creation of `SerializableType` by this method is not a problem but currently it raises exception.

Reply:
The initial coder of all this has seemingly considered a throw was required if another type uses the same NHibernate type name, even in the serializable case. But the `SerializableType.Name` implementation causes all classes implementing `ISerializable` to have the same name, `serializable`.

Otherwise the name is the class full name. This does not make sense for me, but I am unsure what is the right change: cease registering the NHibernate type name for `SerializableType`? Change `SerializableType.Name` implementation? Is registering a type with the name `serializable` useful to anyone?

Reply:
>But the `SerializableType.Name` implementation causes all classes implementing `ISerializable` to have the same name, `serializable`

No. It's "serializable" only for `ISerializable` type. So name is unique for each type - full class name.

It seems the whole point of caching here is to avoid full parsing on each `HeuristicType(typeName)` call. But I don't see as issue creating two different instances of `SerializableType` due to some racing conditions. So I think we should simply replace reference equality with object equality:
`if (added && !Equals(typeByTypeOfName.GetOrAdd(type.Name, type), type))`

Reply:
Actually it seems there is no legit way that two different instances for the same `SerializableType` are stored in dictionary for AQN and FCN. So I guess check is good as it is - sorry for all the fuss.